File: Leetcode/maximum-product-subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int maxProduct(vector<int> &nums)
    {
        int mx = 1;
        int mn = 1;

        int so_far = INT_MIN;
        for (auto &x : nums)
        {
            so_far = max(so_far, x);
        }

        for (auto &x : nums)
        {
            if (x == 0)
            {
                mx = 1;
                mn = 1;
            }
            int old_mx = mx;
            mx = max(x, max(mx * x, mn * x));
            mn = min(x, min(old_mx * x, mn * x));

            so_far = max(so_far, mx);
        }

        return so_far;
    }
};

File: Leetcode/maximum-product-subarray.js
// problem link
// https://leetcode.com/problems/maximum-product-subarray/description/

var maxProduct = function(nums) {
    let mx = 1;
    let mn = 1;
    let so_far = Number.MIN_SAFE_INTEGER;

    for(let num of nums) {
        let mx_pro = mx*num;
        let mn_pro = mn*num;

        mx = Math.max(num, mx_pro, mn_pro);
        mn = Math.min(num, mx_pro, mn_pro);

        so_far = Math.max(so_far, mx);
    }

    return so_far;
};